database/mongo.py

```diff
@@ -1,28 +1,62 @@
-from os import getenv
 from motor.motor_asyncio import AsyncIOMotorClient as MongoClient
-from motor.motor_asyncio import AsyncIOMotorCollection as MongoCollection
-from database import users
+from bson.json_util import loads,dumps
+import aiofiles, aiofiles.os
+from os import getenv
+from discord.ext import tasks, commands as cmds
+from . import users
+
+
+class MongoCog(cmds.Cog, name='Mongo update script'):
+    def __init__(self, bot):
+        self.bot = bot
+        self.mongo_users = MongoClient(getenv('MONGO_URI')).popskill.user_links
+        self.token = None
+        self.watch.start()
+
+    @tasks.loop(count=1)
+    async def watch(self):
+        async with self.mongo_users.watch([{'$match': {'operationType': {'$in': ['insert', 'update', 'replace']}}}],
+                                          full_document='updateLookup',
+                                          resume_after=self.token) as stream:
+            async for change in stream:
+                entry = change['fullDocument']
+                if 'steam_id' in entry:
+                    await users.add_steam64_id(entry['discord_id'], entry['steam_id'])
+                self.token = stream.resume_token
 
+    @watch.before_loop
+    async def get_token(self):
+        try:
+            async with aiofiles.open('resume', 'rb') as tokenfile:
+                self.token = loads(await tokenfile.read())
+        except FileNotFoundError:
+            pass  # token is None by default
 
-async def copy_steam_ids():
-    mongo_client = MongoClient(getenv('MONGO_URI'))
-    mongo_users: MongoCollection = mongo_client.popskill.user_links
-    user_list = await mongo_users.find().to_list(length=None)
-    users_with_steam_ids = []
-    for user in user_list:
-        if 'steam_id' in user.keys():
-            users_with_steam_ids.append(user)
-    await users.add_steam64_ids(users_with_steam_ids)
+    @watch.after_loop
+    async def store_token(self):
+        if self.token is None:
+            return
 
+        async with aiofiles.open('resume', 'wb') as tokenfile:
+            await tokenfile.write(dumps(self.token))
 
-if __name__ == '__main__':
-    import asyncio
-    import aiosqlite
-    from dotenv import load_dotenv
-    load_dotenv("../.env")
+    @watch.error
+    async def log_error(self, error):
+        info = await self.bot.application_info()
+        await info.owner.send(
+            f'**{type(error).__name__}**\n{error}'
```

Comment:
Please could you make this explicit as to where the error came from? It should be clear that it comes from this module, and the exact line/command that raised the error would also be super useful.

Reply:
If I replace this with `logging.error`, we could later make a `LoggingHandler` that DMs you?

Reply:
Is there any way that we could make it raise something that's handled by `on_command_error`? Currently `on_command_error` sends the error message to a log channel and then raises `logging.error`.

Reply:
Since it's not part of a command, no exception can propagate to `on_command_error`. In fact that function wouldn't work here, since it uses `ctx.author`/`ctx.message`/`ctx.channel` which don't make any sense for a long-running internal task.

I think it's best to `logging.error` in both places and use a `LoggingHandler` to send to the log channel? `on_command_error` should probably also tell the user what happened, similar to [what I've done for ratingbot](https://github.com/mxbi/popskill/blob/20ff6675ee1375bcf6b2d857a5aeba269a36012d/discord_app.py#L43).

Reply:
I've replaced with logging.error for now, I'll work on a logging handler for a future PR

Reply:
Yep, sounds like this is the best plan